### PR TITLE
[CARBONDATA-2247][SDK] Support write Index file in CarbonWriter

### PR DESCRIPTION
**Pull request by jackylk**
1. Support write index file in CarbonWriter in SDK
2. Support persist schema in CarbonWriter in SDK
3. Add more testcase in CSVCarbonWriterSuite

 - [X] Any interfaces changed?
 No
 - [X] Any backward compatibility impacted?
 No
 - [X] Document update required?
No
 - [X] Testing done
CSVCarbonWriterSuite
 - [X] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
NA

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2958/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/4203/

**Comment by jackylk** [comment]
retest this please

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3854/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/4212/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2967/

File: core/src/main/java/org/apache/carbondata/core/metadata/schema/table/TableSchemaBuilder.java [hunk 1]
```diff
@@ -42,6 +45,16 @@
 
   private List<ColumnSchema> otherColumns = new LinkedList<>();
 
+  private int blockSize;
+
+  public TableSchemaBuilder blockSize(int blockSize) {
+    if (blockSize == 0) {
```
Comment (xuchuanyin) [inline core/src/main/java/org/apache/carbondata/core/metadata/schema/table/TableSchemaBuilder.java]: Logic and prompt are inconsistent, should be '<='

Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/metadata/schema/table/TableSchemaBuilder.java]: ok

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/4233/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2989/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3875/

**Comment by QiangCai** [comment]
LGTM

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/3018/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/4262/

**Pull request closed without merging**